README
== A quick note

This is a fork of the official repository, plus a couple community-provided
patches to fix compile issues on newer raspberrypi toolchains.


== Overview

mochad is a Linux TCP gateway daemon for the X10 CM15A RF (radio frequency) and
PL (power line) controller and the CM19A RF controller. 

mochad stands for Multiple Online Controllers for Home Automation Daemon.

0.1.21 Fixed many compile errors on Rasberry PI - S Porter steve.porter@yahoo.com
0.1.19 Heard rumblings of this version somewhere out there
0.1.16 Fix "Could not find endpoints" problem.
0.1.15 Add support for 2011 CM19A.
0.1.14 Fix OpenRemote too many open sockets problem. Add shutters/blinds command.
0.1.13 Fix -d debug option. Add support for DS12A.
0.1.12 Add --raw-data option.
0.1.11 Fix xdim (extended dim) that only worked for unit code 3.
0.1.10 Fix major client problem. Commands from clients stopped working.
0.1.9 Add support for OpenRemote 2.0 with iPhone and Android clients.
0.1.8 Add support for Chumby application.


== Build it

mochad compiles and runs on Ubuntu 10.04 and 10.10. The libusb-1.0 development
files are required. To install,

    sudo apt-get install libusb-1.0-0-dev

If you are building on a Beagleboard/Beaglebone (http://beagleboard.org)
using the native compiler, do the following to install libusb-1.0.

    opkg install libusb-1.0-dev

The usual pattern for building the code applies.

    tar xzf mochad-...tar.gz
    cd mochad-...
    ./configure
    make
    sudo make install

Please see http://sourceforge.net/apps/mediawiki/mochad/index.php?title=Main_Page
for details on building for OpenWrt.

== Run it

Plugging a CM15A or CM19A should launch mochad. There no need to run mochad
by hand or run it from /etc/rc.local. mochad is launched by a udev rule that
matches the CM15A and CM19A USB vendor and product ID codes.

== Use it

Connect to mochad using netcat. For example, 

    nc localhost 1099

You can connect to mochad from another host by specifying the IP of the host
running mochad. netcat is available for Windows at
<http://joncraton.org/blog/netcat-for-windows>.

Everything sent by mochad appears on netcat standard output. In the simplest
use case, mochad/netcat can be used to see X10 PL and RF activity. For example,
run mochad on one host with a CM15A then connect to it using netcat from a
netbook. Walk around with various X10 RF remote controls and the netbook to see
which remotes work from various locations.

Sample output:

    12/07 20:49:10 Rx RF HouseUnit: C3 Func: Off
    12/07 20:49:10 Tx PL HouseUnit: C3
    12/07 20:49:10 Tx PL House: C Func: Off
    12/07 20:49:10 Rx RF HouseUnit: C3 Func: Off
    12/07 20:49:10 Tx PL HouseUnit: C3
    12/07 20:49:10 Tx PL House: C Func: Off
    12/07 20:49:17 Rx RFSEC Addr: 0x11 Func: Lights_Off_SH624
    12/07 20:49:17 Rx RFSEC Addr: 0x11 Func: Lights_Off_SH624
    12/07 20:49:37 Rx RFSEC Addr: C6:1B:00 Func: Motion_alert_MS10A
    12/07 20:49:37 Rx RFSEC Addr: C6:1B:00 Func: Motion_alert_MS10A
    12/07 20:49:37 Rx RFSEC Addr: C6:1B:00 Func: Motion_alert_MS10A
    12/07 20:49:37 Rx RFSEC Addr: C6:1B:00 Func: Motion_alert_MS10A

Everything sent to netcat standard input is sent to mochad. The following is a
brief list of mochad commands.

    pl a1 [on|off]
    pl a1 [dim|bright] 0..31
    pl a1 xdim 0..255

    pl a1
    pl a2
    pl a [on|off|dim|bright|xdim|all_lights_on|all_lights_off|all_units_off]

    rf a1 [on|off|dim|bright]

    st  -- show device status including RF security devices

By default, received RF X10 commands are repeated on the PL interface for all
house codes. This can be changed using the rftopl command (RF to PL repeater).

    rftopl *    -- enable all house codes (default)
    rftopl abc  -- enable for house codes A, B, and C. Disable for all others.
    rftopl 0    -- disable for all house codes. This is useful if more than
                    one CM15A is in use. Leaving both enabled will result in 
                    PL collisions and redundant commands.

Please see http://sourceforge.net/apps/mediawiki/mochad/index.php?title=Main_Page
for more details on command and event messages.


If you want to debug X10 RF, or gateway raw frames to another program, you can
use the --raw-data argument.
See apps/mochad.scr for an example on how to gateway frames with misterhouse.

Another use is to simply be able to get all the raw RF frames to see if the
multiple copies received are identical (good), off by a few bits (RF marginal),
or nonsensical (bad).

    07/01 23:34:42 Rx RF HouseUnit: A1 Func: Off
    07/01 23:34:42 Raw data received: 5D 20 60 9F 20 DF 
    07/01 23:34:42 Raw data received: 5D 20 60 9F 20 DF 
    07/01 23:34:43 Raw data received: 5D 20 60 9F 20 DF 
    07/01 23:34:43 Raw data received: 5D 20 60 9F 20 DF 
    07/01 23:34:43 Raw data received: 5D 20 60 9F 20 DF


For examples of controlling shutters and blinds, see the following.

https://sourceforge.net/apps/mediawiki/mochad/index.php?title=Shutter_and_Blinds

== Multiple controllers

The Perl program mochamon.pl shows how to monitor more than one instance of
mochad. It is possible to create a master home automation master which talks to
more than one X10 controller. For example, a mocha master could receive RF
commands from controller A then send PL commands on controller B in a detached
building.

== Command line X10 control

Command line control is useful for controlling devices from cron or CGI programs.

    $ echo "pl a all_lights_off" | nc localhost 1099

More generally, create a file name x10cmd with the following

    echo "$@" | nc localhost 1099

Make the file executable.

    $ chmod +x x10cmd

Run x10cmd with command line parameters to control X10 devices.

    $ ./x10cmd pl a all_lights_off

== Web Interface

The cgi/ sub-directory holds a simple Perl CGI web application that uses
mochad. The  interface is designed for small hand held devices over slow
connections (not broadband) so the web pages are very basic and small.
No GIFs or JPGs are used. To accomodate older devices with limited web browsers,
CSS, Javascript, or AJAX are not used. The web applications has buttons at
the top to send commands to mochad. The remainder shows status information
for three hosts running mochad.

Please see http://sourceforge.net/apps/mediawiki/mochad/index.php?title=Main_Page
for more details.

== Hardware

X10 controllers: CM15A and CM19A

Various X10 devices that have been tested with mochad:
    AM486, LM465 softstart, PAM02, RR501 (2 way), DS10A, MS10A, KR10A, SH624,
    KR19A

NOTE: CM15A macros and timers are NOT supported. For best results, use AHP
(ActiveHome Pro) to clear interface memory and disable the internal RF to PL
repeater. This ensures the CM15A behaves like a transceiver.

== Files

    *.c *.h     -- C source files
    mochamon.pl -- Perl program showing how to connect to three instance of 
                    mochad and dumping out the output to standard out.
    cgi/        -- Sample Perl CGI utilizing more than one mochad

== Acknowledgments

The Perl CGI scripts uses the venerable cgi-lib.pl copyright Steven E. Brenner.
cgi-lib is simpler and smaller than CGI.pm which is appropriate for small
embedded systems such as WiFi routers. 

For OpenWrt based systems, using LuCi would make more sense than Perl since
LuCi is already present in the system. However, the Perl CGI scripts work fine
with lighttpd on OpenWrt.

This project in not in any way affliated with X10 Corp. (www.x10.com). Do
not contact them about this driver since they know nothing about it.

== References

The following documents were useful in development of this program.

X10 Power Line Protocol

    <http://software.x10.com/pub/manuals/xtc798.doc>
    <http://software.x10.com/pub/manuals/xtc798.pdf>
    <http://software.x10.com/pub/manuals/xtdcode.pdf>

X10 RF including RF Security Protocol

    <http://www.edcheung.com/automa/rf.txt>

X10 CM11A Protocol

    <http://software.x10.com/pub/manuals/cm11a_protocol.txt>

Linux Daemon HowTo

    <http://www-theorie.physik.unizh.ch/~dpotter/howto/daemonize>



ChangeLog
mochad-0.1.21

    * 01/24/22: Fixed many compile errors on Rasberry PI - S Porter steve.porter@yahoo.com
    * Added mochad.h file
    * Added mod by J Jones 2018-12-14 here: https://sourceforge.net/p/mochad/discussion/1320002/thread/582c852aa0/
 
mochad-0.1.19

    * Heard rumblings of this version somewhere out there

mochad-0.1.17

    * Fix for Linux distros using systemd-udev such as Raspbian 4.4.
    * Fix for newer versions of autotools.

mochad-0.1.16

    * Fix "Could not find endpoints" problem.

mochad-0.1.15

    * Add support for newer CM19A built in 2011 or later.

mochad-0.1.14

    * Fix OpenRemote too many open sockets problem. This result in delayed
    responses to button presses.
    * Add extended_code_1 command. This can be used to send shutters/blinds
    control commands.

mochad-0.1.13

    * Fix -d debug command line option.

    * The SP554A motion sensor with the switch in the home/away position is
    now correctly decoded.

    * Add exit to potential infinite loops.

    * Add message for DS12A with cover off (tamper). The DS12A looks like
    two DS10As with the cover on.


mochad-0.1.12

    * Add --raw-data option from Marc Merlin. Programs such as MisterHouse
    can decode the raw RF data.

mochad-0.1.11

    * Fix xdim (extended dim) that only worked for unit code 3.

mochad-0.1.10

    * Fix major client socket problem. Commands from clients stopped working.

mochad-0.1.9

    * Add commands getstatus and getstatussec.
    getstatus <HU>          -- Get device HU on/off status
    getstatus <HU> xdim     -- Get device xdim status
    getstatusec <RFSECAddr> -- Get RF security device status such as DS10A.

    * Add port 1101 for OpenRemote 2.0. This is similar to port 1099 but does
    not include unsolicited event messages. OR polls for device events using
    getstatus and getstatussec.

mochad-0.1.8

    * Add support for X10 KR15A Big Red Button.
    * Allow '\r' as well as '\n' for line termination
    * Add more support for Flash XMLSocket clients including cross domain
    policy request and response for Chumby widgets. This is required to allow
    a Chumby widget to talk to mochad through an XMLSocket. Chumby widget
    should be released real soon now.
    * Add python-tk demo GUI application. Supports on and off buttons plus
    brighness slider using LM465 soft-start lamp modules. The slider will not
    work with older LM465 modules which lack soft-start. See apps/x10-tk.py.

mochad-0.1.7

    * Add Flash XMLSocket support on port 1100 for Chumby widgets.
    * Fix RF checksum problem when receiving from CM15A for house codes 9-16.

mochad-0.1.6

    * Add RF event filter to eliminate the flurry of duplicate events.
    This makes it much easier to write applications.
    * Skip extra leading 0x5D on received RF packets from CM15A.

mochad-0.1.5

    * Fix RFCAM command so it shows the transmit message.
    * Fix RF command which sent the wrong house in some cases. [ID: 3197002]

mochad-0.1.4

    * Fix RFCAM command so it takes house code as a parameter. For example,
    RFCAM C CAMUP sends a camera pan up command on house code C.
    Previously, RFCAM commands always used house code N. 

mochad-0.1.3

    * Change the method for sending dim/bright commands so it matches AHP
    The new method uses 3 bytes instead of 2. This seems to work better
    with LM465 soft-start devices. xdim works much better and is easier to use
    but does not work with all lamp modules.
    * Add example client programs in Perl and bash.
    * Add --version command line parameter.
    * Add support for SP554 motion sensor

mochad-0.1.2

    * Fix device name in log messages. CM19A was reported as CM15A.
    * Add install option that makes the controllers hot pluggable. udev in
    Ubunutu and hotplug2 in OpenWrt
    * Convert to use autoconf and automake.
    * Update mochad wiki pages with command and event reference. Also add
    instructions for building with OpenWrt.


decode.h
/*
 * Copyright 2010-2011 Brian Uechi <buasst@gmail.com>
 *
 * This file is part of mochad.
 *
 * mochad is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * mochad is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mochad.  If not, see <http://www.gnu.org/licenses/>.
 */

// ADDED HERE S Porter FROM GLOBAL.H, MADE "extern"
// to stop multiple definition errors 02/04/22
extern int Cm19a;
extern int PollTimeOut;

/* 1 bit per house code, 1=RF to PL, 0=off, default all house codes on */
extern unsigned short RfToPl16;

extern unsigned short RfToRf16;


extern int raw_data;

const char *findSecEventName(unsigned char secev);

const char *findSecRemoteKeyName(unsigned char secev);

int findCamRemoteCommand(const char *keyname);

void cm15a_decode_plc(int fd, unsigned char *buf, size_t len);
    
void cm15a_decode_rf(int fd, unsigned char *buf, unsigned int len);

void cm15a_decode(int fd, unsigned char *buf, unsigned int len);



global.h
/*
 * Copyright 2010-2011 Brian Uechi <buasst@gmail.com>
 *
 * This file is part of mochad.
 *
 * mochad is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * mochad is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mochad.  If not, see <http://www.gnu.org/licenses/>.
 */

struct SecEventRec {
    unsigned char funct;
    const char *name;
};

int Cm19a;
int PollTimeOut;

/* 1 bit per house code, 1=RF to PL, 0=off, default all house codes on */
unsigned short RfToPl16;

unsigned short RfToRf16;

/*  MOVED TO new file mochad.h by S Porter 02/04/22

#define dbprintf(fmt, ...) _dbprintf(fmt, __FILE__,__LINE__, ## __VA_ARGS__)
int _dbprintf(const char *fmt, ...);

// int write_usb(unsigned char *buf, size_t len);

// S Porter: added mod by J Jones 2018-12-14 here: https://sourceforge.net/p/mochad/discussion/1320002/thread/582c852aa0/
int write_usb(unsigned char *buf, int len);     

int statusprintf(int fd, const char *fmt, ...);
int sockprintf(int fd, const char *fmt, ...);

void hexdump(void *p, size_t len);

void sockhexdump(int fd, void *p, size_t len);

void mh_sockhexdump(int fd, void *p, size_t len);

int or20client(int fd);

int del_client(int fd);

*/


mochad.h
/*
 * Copyright 2010-2011 Brian Uechi <buasst@gmail.com>
 *
 * This file is part of mochad.
 *
 * mochad is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * mochad is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mochad.  If not, see <http://www.gnu.org/licenses/>.
 */

// THIS FILE ADDED BY S Porter 02/04/22 TO CLEAN UP COMPILE ERRORS

#define dbprintf(fmt, ...) _dbprintf(fmt, __FILE__,__LINE__, ## __VA_ARGS__)
int _dbprintf(const char *fmt, ...);

// int write_usb(unsigned char *buf, size_t len)

// S Porter: added mod by J Jones 2018-12-14 here: https://sourceforge.net/p/mochad/discussion/1320002/thread/582c852aa0/
int write_usb(unsigned char *buf, int len);

int statusprintf(int fd, const char *fmt, ...);
int sockprintf(int fd, const char *fmt, ...);

void hexdump(void *p, size_t len);

void sockhexdump(int fd, void *p, size_t len);

void mh_sockhexdump(int fd, void *p, size_t len);

int or20client(int fd);

int del_client(int fd);


x10_write.c
/*
 * Copyright 2010-2011 Brian Uechi <buasst@gmail.com>
 *
 * This file is part of mochad.
 *
 * mochad is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * mochad is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mochad.  If not, see <http://www.gnu.org/licenses/>.
 */

/* CM15A output FIFO
 * X10 I/O is very slow so store up pending output then wait for X10ACK
 * (0x55) before sending the next item from the queue.
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdarg.h>
#include <stdlib.h>
// #include "global.h"   // S Porter TO PREVENT DUPLICATE DECLARATION  ERRORS 02/04/22
#include "x10_write.h"

#include "mochad.h"      // ADDED 02/04/22 S Porter

extern int PollTimeOut;  // S Porter ADDED 02/04/22

typedef struct x10out {
    size_t outlen;
    unsigned char outdata[8];
} x10out_t;

static x10out_t Outrecs[256];
#define OUTPTRSSIZE             (sizeof(Outrecs)/sizeof(Outrecs[0]))
static int Outhead = 0;
static int Outtail = 0;
static int Outbusy = 0;

static int next_index(int idx)
{
    return ((idx + 1) % OUTPTRSSIZE);
}

static int add_x10out(unsigned char *buf, size_t buflen)
{
    int nxt;
    x10out_t *nxtrec;

    dbprintf("len %lu\n", buflen);
    if ((nxt = next_index(Outtail)) == Outhead) {
        dbprintf("Outptrs full Outhead/tail %d/%d\n", Outhead, Outtail);
        return -1;
    }

    nxtrec = &Outrecs[nxt];
    nxtrec->outlen = buflen;
    memcpy(nxtrec->outdata, buf, buflen);
    Outtail = nxt;
    return buflen;
}

int send_next_x10out(void)
{
    x10out_t *outrec;

    if (Outbusy) {
        dbprintf("Outhead Outtail %d/%d\n", Outhead, Outtail);
        if (Outhead == Outtail) {
            /* Empty */
            Outbusy = 0;
            PollTimeOut = -1;
        }
        else {
            Outhead = next_index(Outhead);
            outrec = &Outrecs[Outhead];
            write_usb(outrec->outdata, outrec->outlen);
        }
    }
    return 0;
}

int x10_write(unsigned char *buf, size_t buflen)
{
    dbprintf("Outbusy=%d\n", Outbusy);
    if (Outbusy) {
        add_x10out(buf, buflen);
    }
    else {
        Outbusy = 1;
        PollTimeOut = 2*1000;   /* 2 seconds */
        write_usb(buf, buflen);
    }
    return buflen;
}


x10state.c
/*
 * Copyright 2010-2011 Brian Uechi <buasst@gmail.com>
 *
 * This file is part of mochad.
 *
 * mochad is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * mochad is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mochad.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdarg.h>
#include <stdlib.h>
#include <time.h>
// #include "global.h"   // S Porter TO PREVENT DUPLICATE DECLARATION ERRORS 02/04/22
#include "x10state.h"
#include "decode.h"

#include "mochad.h"         // ADDED 02/04/22 S Porter

/* 16 house codes and 16 unit codes = 256 devices
 * For normal (non-security) X10 devices.
 */
typedef unsigned char houseunitaddrs_t[16][16];

/* For X10 security sensors such as MS10A motion sensor and DS10A 
 * door/window sensor
 */
typedef struct _x10secsensor {
    unsigned long secaddr;
    time_t        lastupdate;
    unsigned char secaddr8;     /* 0 17 bit addr, 1 8 bit addr */
    unsigned char sensorstatus;
} x10secsensor_t;

static x10secsensor_t X10sensors[512];
static unsigned int  X10sensorcount = 0;

/* 0x00 = unknown, '1' = ON, '0' = OFF */
static houseunitaddrs_t HouseUnitState;
/* 0x00 = not selected, '1' = selected */
static houseunitaddrs_t HouseUnitSelected;
/* 0..255 dim as set by xdim command */
static houseunitaddrs_t HouseUnitDim;
/* house/unit/func state per house code
 * 0 = house/unit code, 1 = house/function */
static int X10protostate[16];

/* HouseUnitSelect and X10protostate are needed because of the following case.
 * pl b3    # Select b3
 * pl b4    # Select b4
 * pl b on  # Turn on b3 and b4
 */

void hua_sec_init(void) 
{
    memset(HouseUnitState, 0, sizeof(HouseUnitState));
    memset(HouseUnitSelected, 0, sizeof(HouseUnitSelected));
    memset(HouseUnitDim, 0, sizeof(HouseUnitDim));
    memset(X10protostate, 0, sizeof(X10protostate));

    memset(X10sensors, 0, sizeof(X10sensors));
    X10sensorcount = 0;
}

#define issecfunc(x) (((x & 0xF0) == 0x80) || ((x & 0xF0) == 0x00))

/* Remember RF security event */
void hua_sec_event(unsigned char *secaddr, unsigned int funcint, 
        unsigned int secaddr8)
{
    int i;
    unsigned long secaddr32;
    x10secsensor_t *sen;

    secaddr32 = (secaddr[0] << 16) | (secaddr[1] << 8) | secaddr[2] ;
    /* dbprintf("secaddr32 %X func %X issecfunc %d\n", 
            secaddr32, funcint, issecfunc(funcint)); */

    for (i = 0; i < X10sensorcount; i++) {
        sen = &X10sensors[i];
        if (secaddr32 == sen->secaddr) {
            sen->sensorstatus = funcint;
            //if ((funcint & (1<<7)) == 0) {
                sen->lastupdate = time(NULL);
            //}
            return;
        }
    }

    /* Add new device */
    if ((X10sensorcount + 1) >= (sizeof(X10sensors)/sizeof(x10secsensor_t)))
        return;
//    if (issecfunc(funcint)) {
        sen = &X10sensors[X10sensorcount];
        sen->secaddr = secaddr32;
        sen->secaddr8 = secaddr8;
        sen->sensorstatus = funcint;
        sen->lastupdate = time(NULL);
        X10sensorcount++;
//    }
}

static void hua_dbprint(void)
{
    int h, u;
    char buf[4096];

    strcpy(buf, "Selected:");
    for (h = 0; h < 16; h++) {
        for (u = 0; u < 16; u++) {
            if (HouseUnitSelected[h][u])
                snprintf(buf+strlen(buf), sizeof(buf)-strlen(buf), "%c%d,", h+'A', u+1);
        }
    }
    strcat(buf, "\n");
    dbprintf(buf);

    strcpy(buf, "State: ");
    for (h = 0; h < 16; h++) {
        for (u = 0; u < 16; u++) {
            if (HouseUnitState[h][u])
                snprintf(buf+strlen(buf), sizeof(buf)-strlen(buf), "%c%d %c,", h+'A', u+1, 
                        HouseUnitState[h][u]);
        }
    }
    strcat(buf, "\n");
    dbprintf(buf);
}

static void hua_init(int house)
{
    memset(&HouseUnitSelected[house], 0, sizeof(HouseUnitSelected)/16);
}

unsigned char hua_getstatus(int house, int unit)
{
    return HouseUnitState[house][unit];
}

unsigned char hua_getstatus_xdim(int house, int unit)
{
    return HouseUnitDim[house][unit];
}

void hua_setstatus_xdim(int house, int unit, int xdim)
{
    /* dbprintf("%s(%d,%d)\n", __func__, house, xdim); */
    hua_init(house);
    HouseUnitSelected[house][unit] = '1';
    X10protostate[house] = 1;
    HouseUnitDim[house][unit] = xdim;
    if (xdim > 0)
        HouseUnitState[house][unit] = '1';
    else
        HouseUnitState[house][unit] = '0';
}

void hua_add(int house, int unit)
{
    switch (X10protostate[house]) {
        case 0:
            HouseUnitSelected[house][unit] = '1';
            break;
        case 1:
            hua_init(house);
            HouseUnitSelected[house][unit] = '1';
            X10protostate[house] = 0;
            break;
        default:
            dbprintf("Invalid state\n");
    }
    // hua_dbprint();
}

static void hua_func_all(int house, unsigned char func)
{
    int u;

    X10protostate[house] = 1;
    for (u = 0; u < 16; u++) {
        HouseUnitState[house][u] = func;
        if (func == '1')
            HouseUnitDim[house][u] = 63;
        else
            HouseUnitDim[house][u] = 0;
    }
    // hua_dbprint();
}

void hua_func_all_on(int house)
{
    hua_func_all(house, '1');
}

void hua_func_all_off(int house)
{
    hua_func_all(house, '0');
}

static void hua_func(int house, unsigned char func)
{
    int u;

    /* dbprintf("%s(%d,%d)\n", __func__, house, func); */
    X10protostate[house] = 1;
    for (u = 0; u < 16; u++) {
        if (HouseUnitSelected[house][u]) {
            HouseUnitState[house][u] = func;
            if (func == '1')
                HouseUnitDim[house][u] = 63;
            else
                HouseUnitDim[house][u] = 0;
        }
    }
    // hua_dbprint();
    /* dbprintf("%s exit\n", __func__); */
}

void hua_func_on(int house)
{
    hua_func(house, '1');
}

void hua_func_off(int house)
{
    hua_func(house, '0');
}

int hua_getstatus_sec(int rf8bitaddr, unsigned long rfaddr)
{
    x10secsensor_t *sen;
    int sensor;

    dbprintf("hua_getstatus_sec(%d,%X)\n", rf8bitaddr, rfaddr);
    for (sensor = 0; sensor < X10sensorcount; sensor++) {
        sen = &X10sensors[sensor];

        dbprintf("hua_getstatus_sec addr8 %d addr %X status %X\n",
                sen->secaddr8, sen->secaddr, sen->sensorstatus);
        if (rf8bitaddr && sen->secaddr8 && (rfaddr == sen->secaddr)) {
            switch (sen->sensorstatus) {
                case 0x04: return 1;    /* alert */
                case 0x84: return 0;    /* normal */
            }
        }
        else if (!rf8bitaddr && !sen->secaddr8 && (rfaddr == sen->secaddr)) {
            switch (sen->sensorstatus) {
                case 0x00:
                case 0x01:
                case 0x04:
                case 0x05:
                case 0x0C:
                case 0x0D:
                    return 1;    /* alert */
                case 0x80:
                case 0x81:
                case 0x84:
                case 0x85:
                case 0x8C:
                case 0x8D:
                    return 0;    /* normal */
            }
        }
    }
    return -1;
}

static int cmpX10sensor(const void *e1, const void *e2)
{
    const x10secsensor_t *sen1 = e1, *sen2 = e2;

    return sen1->secaddr - sen2->secaddr;
}

void hua_show(int fd)
{
    int h, u;
    char buf[2048];
    int labeldone;
    int sensor;
    x10secsensor_t *sen;

    sockprintf(fd, "Device selected\n");
    for (h = 0; h < 16; h++) {
        labeldone = 0;
        buf[0] = '\0';
        for (u = 0; u < 16; u++) {
            if (HouseUnitSelected[h][u]) {
                if (!labeldone) {
                    labeldone = 1;
                    snprintf(buf+strlen(buf), sizeof(buf)-strlen(buf), 
                            "House %c: %d", h+'A', u+1);
                }
                else 
                    snprintf(buf+strlen(buf), sizeof(buf)-strlen(buf), 
                            ",%d", u+1);
            }
        }
        if (labeldone) {
            strcat(buf, "\n");
            sockprintf(fd, buf, strlen(buf));
        }
    }
    
    sockprintf(fd, "Device status\n");
    for (h = 0; h < 16; h++) {
        labeldone = 0;
        buf[0] = '\0';
        for (u = 0; u < 16; u++) {
            if (HouseUnitState[h][u]) {
                if (!labeldone) {
                    labeldone = 1;
                    snprintf(buf+strlen(buf), sizeof(buf)-strlen(buf), 
                            "House %c: %d=%c", h+'A', u+1, HouseUnitState[h][u]);
                }
                else
                    snprintf(buf+strlen(buf), sizeof(buf)-strlen(buf), 
                            ",%d=%c", u+1, HouseUnitState[h][u]);
            }
        }
        if (labeldone) {
            strcat(buf, "\n");
            sockprintf(fd, buf, strlen(buf));
        }
    }
    sockprintf(fd, "Security sensor status\n");
    qsort(X10sensors, X10sensorcount, sizeof(X10sensors[0]), cmpX10sensor);
    for (sensor = 0; sensor < X10sensorcount; sensor++) {
        time_t deltat, mins;
        const char *message;

        sen = &X10sensors[sensor];
        deltat = time(NULL) - sen->lastupdate;
        mins = deltat / 60;
        deltat = deltat - (mins * 60);
        if (sen->secaddr8)
            message = findSecRemoteKeyName(sen->sensorstatus);
        else
            message = findSecEventName(sen->sensorstatus);
        sockprintf(fd, "Sensor addr: %06X Last: %02d:%02d %s \n", sen->secaddr,
                mins, deltat, message);
    }

    sockprintf(fd, "End status\n");
}
